### PR TITLE
 feat: add email confirmation and password reset flow

### LINKED ISSUE #198: Email Confirmation & Password Reset
We need to verify user emails on signup and allow password recovery.  
This ticket adds confirmation tokens, mailers, and “reset password” flows.

---

## Why it matters 💡
| Goal | Benefit |
|------|---------|
| Verify ownership of email | Prevent spam / typos in signup flow |
| Reset forgotten passwords | Reduce support burden; no admin intervention |
| Harden auth | Users must confirm before accessing protected areas |

---

### 1  DB migrations

1. **Create `email_confirmations` table**  
   ```bash
   rails g migration CreateEmailConfirmations 
     user:references token:string:index confirmed_at:datetime sent_at:datetime
   rails db:migrate
   ```

2. **Add `password_reset_token`, `password_reset_sent_at` to `users`**  
   ```bash
   rails g migration AddPasswordResetToUsers 
     password_reset_token:string:index 
     password_reset_sent_at:datetime
   rails db:migrate
   ```

### 2  Model logic

- **User model**  
  ```ruby
  has_one :email_confirmation, dependent: :destroy

  def generate_password_reset!
    update!(password_reset_token: SecureRandom.hex(20),
            password_reset_sent_at: Time.current)
  end

  def clear_password_reset!
    update!(password_reset_token: nil, password_reset_sent_at: nil)
  end
  ```

- **EmailConfirmation model**  
  ```ruby
  belongs_to :user
  before_create { self.token = SecureRandom.hex(20) }
  ```

### 3  Mailers

```bash
rails g mailer UserMailer
```

- **confirmation_email(user, token)** → link `/confirm?token=...`  
- **password_reset_email(user)** → link `/reset_password?token=...`

Configure development to use **Letter Opener** (`config/environments/development.rb`):

```ruby
config.action_mailer.delivery_method = :letter_opener
config.action_mailer.perform_deliveries = true
```

### 4  Controller endpoints

| HTTP | Path | Purpose |
|------|------|---------|
| POST | `/api/v1/confirmations` | resend confirmation |
| GET  | `/api/v1/confirm`       | confirm via token |
| POST | `/api/v1/password_resets` | request reset (send email) |
| PATCH| `/api/v1/password_resets/:token` | set new password |

Add specs for success / invalid token / expired token.

### 5  Routing

```ruby
namespace :api do
  namespace :v1 do
    resources :confirmations, only: [:create]
    get 'confirm', to: 'confirmations#confirm'
    resources :password_resets, only: [:create, :update], param: :token
  end
end
```

### 6  Frontend stubs (optional)

- Simple React screens: “Check your inbox”, “Enter new password”.

---

## ✅ Definition of Done

- [ ] DB migrations created & migrated  
- [ ] `EmailConfirmation` model with token generation  
- [ ] `UserMailer` sends confirmation & reset emails (Letter Opener in dev)  
- [ ] Endpoints: resend confirmation, confirm, request reset, reset password  
- [ ] Specs for all happy/edge cases (tokens invalid, expired, reused)  
- [ ] Manual test: signup → email → confirm → login works  
- [ ] Manual test: request reset → email → change password → login with new password

### PR DESCRIPTION
**Pull request by JossWritesCode**
resolves #198 and fixes bug in seeds

image

### CODE REVIEW

File: backend/app/models/user.rb [hunk 1]
```diff
@@ -42,41 +59,53 @@ def restore
...
-  has_many :subscriptions, through: :user_subscriptions
+  has_many :subscriptions,      through: :user_subscriptions
+
+  # -------------------------------------------------------------------------
+  # Soft-delete helpers
+  # -------------------------------------------------------------------------
+  scope :active, -> { where(deleted_at: nil) }
+
+  def restore
+    update!(deleted_at: nil)
```
[the remark this answers is not in the document]
Comment (JossWritesCode) [reply backend/app/models/user.rb:97]: i'm ok with it with the bang
